dataRetriever/dataPool/proofsCache/proofsCache.go

```diff
@@ -43,39 +45,66 @@ func (pc *proofsCache) addProof(proof data.HeaderProofHandler) {
 		return
 	}
 
-	pc.mutProofsCache.Lock()
-	defer pc.mutProofsCache.Unlock()
+	pc.insertProofByNonce(proof)
 
-	pc.proofsByNonce = append(pc.proofsByNonce, &proofNonceMapping{
-		headerHash: string(proof.GetHeaderHash()),
-		nonce:      proof.GetHeaderNonce(),
-	})
+	pc.mutProofsByHash.Lock()
+	pc.proofsByHash[string(proof.GetHeaderHash())] = proof
+	pc.mutProofsByHash.Unlock()
```

Comment:
Having these two locks separated does not seem to be an optimization (at least, not at a first glance). Since:

```
lock A
do ~trivial (& with short duration) logic around insertion in bucket
unlock A
lock B
do ~trivial (& with short duration) operation of adding a map entry
unlock B
```

Is not necessarily better than:

```
lock C
do ~trivial (& with short duration) logic around insertion in bucket
do ~trivial (& with short duration) operation of adding a map entry
unlock C
```

Reply:
right, updated to keep only one mutex